src/bstr.rs

```diff
@@ -55,6 +55,18 @@ impl BStr {
         unsafe { mem::transmute(slice) }
     }
 
+    #[inline]
+    #[cfg(feature = "std")]
+    pub(crate) fn from_boxed_bytes(slice: Box<[u8]>) -> Box<BStr> {
+        unsafe { mem::transmute(slice) }
```

Comment:
This is unsound, but it can be implemented safely as `Box::from_raw(Box::into_raw(slice) as *mut BStr)`.

Reply:
I think `Box<T>` is guaranteed to have the same layout as `ptr::NonNull<T>` (and `as` casts on pointers (with the same fat metadata) is defined to just be a transmute), so I _think_ transmuting `Box` is ok, but I'm not 100% sure. The module docs only explicitly call out the guarantee when `T: Sized`, because that's the only time when `ptr::NonNull<T>` is layout-compatible with C `T*`.

All that said, `from_raw(into_raw(_) as _)` is definitely sound and more explicit on intent, so it's better to implement it that way. (NB: it _is_ required to do that for e.g. `Arc`, it's just `Box` that has special layout guarantees here.)

Reply:
You may be right. I haven't found any documentation explicitly saying it's safe or unsafe for unsized types. And since `transmute` is used so often for `Box`, I doubt the behavior will ever change

That said, I still prefer `from_raw` and `into_raw`, but keeping the transmutes is probably okay

Reply:
I'll update the PR tomorrow, `from_raw(into_raw(_) as _))` is just the better way of writing this, anyway.